src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -326,21 +355,28 @@ private void CompactLRU()
         /// </summary>
         private void Compact()
         {
-            var newCacheSize = CalculateNewCacheSize();
-            while (_map.Count > newCacheSize)
+            try
             {
-                // Since all items could have been removed by the public TryRemove() method, leaving the map empty, we need to check if a default value is returned.
-                // Remove the item from the map only if the returned item is NOT default value.
-                var item = _map.FirstOrDefault();
-                if (!item.Equals(default))
+                int newCacheSize = CalculateNewCacheSize();
+                while (_map.Count > newCacheSize)
                 {
-                    if (_map.TryRemove(item.Key, out var cacheItem))
-                        OnItemRemoved?.Invoke(cacheItem.Value);
+                    // Since all items could have been removed by the public TryRemove() method, leaving the map empty, we need to check if a default value is returned.
+                    // Remove the item from the map only if the returned item is NOT default value.
+                    KeyValuePair<TKey, LRUCacheItem<TKey, TValue>> item = _map.FirstOrDefault();
+                    if (!item.Equals(default))
+                    {
+                        if (_map.TryRemove(item.Key, out LRUCacheItem<TKey, TValue> cacheItem))
+                        {
+                            OnItemCompacted?.Invoke(cacheItem.Value);
+                            _compactedItems.Add(cacheItem);
```

Comment:
I don't quite understand, why use the compacted cache here. Why not just dispose/release the signature provider immidiately?